src/ParallelAlgorithms/NonlinearSolver/NewtonRaphson/ElementActions.hpp

```diff
@@ -0,0 +1,470 @@
+      const Parallel::GlobalCache<Metavariables>& /*cache*/,
+      const ArrayIndex& /*array_index*/, const ActionList /*mete*/,
+      const ParallelComponent* const /*meta*/) noexcept {
+    // Retrieve reduction data from inbox
+    auto globalization_result = std::move(
+        tuples::get<Tags::GlobalizationResult<OptionsGroup>>(inboxes)
+            .extract(db::get<Convergence::Tags::IterationId<OptionsGroup>>(box))
+            .mapped());
+    ASSERT(
+        std::holds_alternative<Convergence::HasConverged>(globalization_result),
+        "No globalization should occur for the initial residual");
+    auto& has_converged = get<Convergence::HasConverged>(globalization_result);
+
+    db::mutate<Convergence::Tags::HasConverged<OptionsGroup>>(
+        make_not_null(&box),
+        [&has_converged](const gsl::not_null<Convergence::HasConverged*>
+                             local_has_converged) noexcept {
+          *local_has_converged = std::move(has_converged);
+        });
+
+    // Skip steps entirely if the solve has already converged
+    constexpr size_t complete_step_index =
+        tmpl::index_of<ActionList,
+                       Globalize<FieldsTag, OptionsGroup, Label>>::value +
+        1;
+    constexpr size_t this_action_index =
+        tmpl::index_of<ActionList, InitializeHasConverged>::value;
+    return {std::move(box), false,
+            get<Convergence::Tags::HasConverged<OptionsGroup>>(box)
+                ? complete_step_index
+                : (this_action_index + 1)};
+  }
+};
+
+template <typename FieldsTag, typename OptionsGroup, typename Label>
+struct PrepareStep {
+ private:
+  using fields_tag = FieldsTag;
+  using correction_tag =
+      db::add_tag_prefix<NonlinearSolver::Tags::Correction, fields_tag>;
+  using linear_operator_applied_to_correction_tag =
+      db::add_tag_prefix<LinearSolver::Tags::OperatorAppliedTo, correction_tag>;
+  using globalization_fields_tag =
+      db::add_tag_prefix<NonlinearSolver::Tags::Globalization, fields_tag>;
+
+ public:
+  using const_global_cache_tags =
+      tmpl::list<NonlinearSolver::Tags::DampingFactor<OptionsGroup>,
+                 logging::Tags::Verbosity<OptionsGroup>>;
+
+  template <typename DbTagsList, typename... InboxTags, typename Metavariables,
+            typename ArrayIndex, typename ActionList,
+            typename ParallelComponent>
+  static std::tuple<db::DataBox<DbTagsList>&&> apply(
+      db::DataBox<DbTagsList>& box,
+      const tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+      const Parallel::GlobalCache<Metavariables>& /*cache*/,
+      const ArrayIndex& array_index, const ActionList /*meta*/,
+      const ParallelComponent* const /*meta*/) noexcept {
+    if (UNLIKELY(get<logging::Tags::Verbosity<OptionsGroup>>(box) >=
+                 ::Verbosity::Debug)) {
+      Parallel::printf(
+          "%s " + Options::name<OptionsGroup>() + "(%zu): Prepare step\n",
```

Comment:
I'm not sure what `array_index` is in the elliptic code. Is it the same ElementId as for evolutions? If so, it will not print nicely as a size_t (it is a few ints long), and I've had best results with printing it as a string from, e.g., `MakeString{} << array_index`.

Reply:
It is printed as a string! It corresponds to the `%s` at the start of the line :)

Reply:
Oh no, I can't read!